Add Mixin guides (#133)
FabricMC/fabric-docs: develop/mixins/introduction.md

Reviewer: File name should be `index.md`

Then, we can annotate our `logAgeAccess` method with our annotation in the mixin class.

```java

@Mixin(Person.class)
public abstract class PersonMixin {
    private int phoneNumber;

    public void sayHello() {
        System.out.println("Hello Fabric!");
    }

    @Inject(method = "getAge", at = @At("HEAD"))
    public void logAgeAccess() {
        System.out.println("The age field of person " + this + " has been accessed!");
    }
}

```

The `at` parameter here takes in an `@At` annotation as an argument, which allows us to specify with detail where we want to inject our method into. Later in this guide we will explore how to use this annotation in more detail.

Now, the `Person` class will look like this at runtime:

```java
public class Person {
    private final String name;
    private final int age;
    private int mixinMerged_phoneNumber;

    public Person(String name, int age) {
        this.name = name;
        this.age = age;
    }

    public String getName() {
        return this.name;
    }

    public int getAge() {

>         mixinMerged_logAgeAccess(); // <- Injected call to mixinMerged_logAccess

Reviewer: ```suggestion
        mixinMerged_logAgeAccess(); // ![code highlight]
```

        return this.age;
    }

    public void mixinMerged_sayHello() {
        System.out.println("Hello Fabric!");
    }

    @Inject(method = "getAge", at = @At("HEAD"))
    public void mixinMerged_logAgeAccess() {
        System.out.println("The age field of a person has been accessed!");
    }
}
```

These types of annotations, that allow you to place calls to methods inside already existing methods of some target class, are called **injector annotations**, or just **injectors**, and they are the main tool offered by Mixin to help you modify Minecraft's source code.

There are many different types of annotations, many of them can do more than inject simple calls to method, and they each have their purpose and use cases. Later we will learn how to choose the right one for the goal and how to use them appropriately.

> ## Wait, What Do You Mean By "At Runtime"? {#wait-what-do-you-mean-by-at-runtime}

Reviewer: ```suggestion
## Wait, What Do You Mean by "At Runtime"? {#wait-what-do-you-mean-by-at-runtime}
```

Suppose that in our `logAgeAccess()` injected method, we would also like to show the age of the person.

We can attempt to modify our injected method like so:

```java
@Inject(method = "getAge", at = @At("HEAD"))
public void mixinMerged_logAgeAccess() {
    System.out.println("The age field with value " + this.age + "of a person has been accessed!");
}
```

But there's a problem. Remember that before **runtime**, the mixin class just looks like a regular class to the compiler.

When we type `this.age`, you are trying to access the `age` field of the `Person` class, which is the target class, but it's not the actual class that you are writing it into. The compiler thinks you are trying to access the `age` field of the `PersonMixin` class.

If `age` is a field that doesn't exist in the mixin class, then attempting to run the game will result in a compile time error.

But here's the thing, we _know_ that the `age` field will be available to the injected method when the game runs, as the method will be merged with the `Person` class. If only there was a way to tell the compiler that `age` will exist and to let us access it.

> Let's consider an annotation that does that, called `@Shadow`. What we do now is that we add the `age` field to our mixin class like any other regular field, but we annotate it with `@Shadow` to let the compiler know that this field will refer to the `age` field of the mixin's target class.

Reviewer: ```suggestion
There is an annotation that does just that, called `@Shadow`. What we need to do is to add the `age` field to our mixin class, like any other regular field, but we annotate it with `@Shadow` to let the compiler know that this field will refer to the `age` field of the mixin's target class.
```

> Then, we can modify our mixin like so:
> 

Reviewer: ```suggestion
```

```java
@Mixin(Person.class)
public abstract class PersonMixin {
    @Shadow
    private int age;
    private int phoneNumber;

    public void sayHello() {
        System.out.println("Hello Fabric!");
    }

    @Inject(method = "getAge", at = @At("HEAD"))
    public void logAgeAccess() {
        System.out.println("The age field with value " + this.age + "of a person has been accessed!");
    }
}
```

Now, we will not get a compiler error when we try to run the game. At runtime, `this.age` will refer to the `age` field of the `Person` class and, if such a field does in fact exist in the target class, then we will not get a runtime error either.

What we just did, the process of accessing a member of the target class by marking it as "will exist" in our mixin class, is called **Shadowing**. In the case of methods, the method declaration can be left with an empty body, if the mixin class is non-abstract, or abstract of the mixin class is abstract.

> On the topic of abstract mixin classes, let's finally discuss why mixin classes should be abstract.
> 

Reviewer: ```suggestion
```

## Abstract Mixin Classes {#abstract-mixin-classes}

> There are a few advantages of making your mixin classes abstract. Apart from making semantic sense (a mixin class should never be instantiated directly), it allows you to access the `this` instance.

Reviewer: ```suggestion
Let's finally discuss why mixin classes should be abstract. There are a few advantages of making your mixin classes abstract. Apart from making sense semantically (a mixin class should never be instantiated directly), it allows you to access the `this` instance.
```

```java
    @Inject(method = "getAge", at = @At("HEAD"))
    public void mixinMerged_logAgeAccess() {
        System.out.println("The age field with value " + ((Person) (Object) this).age + "of a person has been accessed!");
    }
```

Note that, for passing the `this` instance to methods, you still need to cast it, first to `Object`, then to the target class, explicitly.

## Wrapping Up {#wrapping-up}

We have now learned the key concepts needed to understand mixins, which will help us understand the different tools offered by mixins, and certain "tricks" you can accomplish with them:

- A **Mixin class** is a Java class that declares changes and additions to a target class.
- The members of a mixin class are merged **at runtime** with the target class. Before this, certain "hacks" are needed to circumnavigate the compiler.
- A Mixin class supports injecting calls to merged methods inside other methods of the target class.
- **At runtime** the mixin class no longer exists. Thus, it should never be referenced directly from anywhere in your mod.

> 
> We are now ready to get started with creating our first mixin in our Fabric mod!

Reviewer: ```suggestion
```
